### PR TITLE
New example, improvements, and readme update

### PR DESCRIPTION
**Pull request by thoov**
I am still working on a few CSS tweaks so dont merge until they are in

- [x] Improve README
- [x] Clean up some handlebars from 'email' to 'todo'
- [x] New example
- [x] Move key to be string and not an object

### CODE REVIEW

File: addon/state-for.js [hunk 1]
```diff
@@ -22,8 +23,26 @@ function createStates(container, stateName) {
...
 
-  states[stateName] = new Ember.MapWithDefault({
-    defaultValue: () => buildDefaultState.call(this, StateFactory)
+  /*
+  * Overriding the default get function to support the defaultValue function
+  * recieving a custom context.
+  * Taken from:
+  * https://github.com/emberjs/ember.js/blob/6a9c789295/packages/ember-metal/lib/map.js#L467
+  */
+  MapWithDefault.prototype.get = function(key, context) {
```
Comment (thoov) [inline addon/state-for.js]: @stefanpenner not sure if this is the best way of accomplishing this

File: bower.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "ember-state-services",
   "dependencies": {
-    "ember": "1.13.8",
+    "ember": "2.0.1",
```
Comment (thoov) [inline bower.json]: Im going to revert this

File: tests/dummy/app/templates/todos/new.hbs [hunk 3]
```diff
@@ -0,0 +1 @@
+laksdjl
```
Comment (thoov) [inline tests/dummy/app/templates/todos/new.hbs]: I will be updating this

File: README.md [hunk 4]
```diff
@@ -67,14 +57,42 @@ export default Ember.Component.extend({
...
+  }
+});
+
+export default MyStateObject;
+```
+
+### Using ember-buffered-proxy
+
+```js
+import BufferedProxy from 'ember-buffered-proxy/proxy';
```
Comment (stefanpenner) [inline README.md]: export { default } ...

File: addon/state-for.js (hunk 1 above)
Comment (stefanpenner) [inline addon/state-for.js]: why not just use `bind` ?

Comment (thoov) [inline addon/state-for.js]: a simple bind wont work in this case:

```js

// components/foo.js
export default Ember.Component.extend({
  data: stateFor('email', {key: 'somekey'})
});


//components/bar.js
export default Ember.Component.extend({
  data: stateFor('email', {key: 'somekey'})
});
```

If 2 components use the same state then each component's stateFor calls:

```js
states[stateName].get(this.get(key))
```

However, defaultValues context will be 'binded' to the first component and will not be 'dynamic'

```js

new MapWithDefault({
  defaultValue: function() { // how do we make the context here dynamic to each component }
});
```

File: addon/custom-map-with-default.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import Ember from 'ember';
+
+class CustomMapWithDefault extends Ember.MapWithDefault {
```
Comment (thoov) [inline addon/custom-map-with-default.js]: @stefanpenner I'm extending MapWithDefault this way now instead of just highjacking the prototype directly

**Comment by thoov** [comment]
@stefanpenner r?

File: addon/custom-map-with-default.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  constructor(options) {
```
Comment (stefanpenner) [inline addon/custom-map-with-default.js]: I think the default constructor could just be inherited, without need for shadowing in this subclass

File: tests/dummy/app/components/routable-todo-message.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import Ember from 'ember';
+import stateFor from 'ember-state-services/state-for';
+
+export default Ember.Component.extend({
+  data: stateFor('todos', { key: 'todo.id' }),
```
Comment (stefanpenner) [inline tests/dummy/app/components/routable-todo-message.js]: here it is `key: 'todo.id' above it is merely the key value, are both supported ?

**Comment by thoov** [comment]
@stefanpenner What should the key be for the weakmap? Every key that I can think of doesnt work. IE what should this be 

```js
myWeakMap.set(XXX, state);
```

File: tests/dummy/app/components/routable-todo-message.js (hunk 7 above)
Comment (stefanpenner) [inline tests/dummy/app/components/routable-todo-message.js]: this should just be `todo`

**Comment by stefanpenner** [comment]
> But then state is this mapWithDefault object and that doesn't make since anymore since there would only ever be one key inside of it (seems like overkill?)

sorry I don't quite understand what you are saying

**Comment by thoov** [comment]
Here is a sample object:
```js
Ember.Component.extend({
  email: { id: 1, subject: 'Foo', message: 'Hello World' },

  data: stateFor('email-draft', 'email'),
  otherData: stateFor('email-upload', 'email')
});
```

Internally this would create an object like this:

```js
{
  'email-draft': WeakMapA,
  'email-upload': WeakMapB 
}
```

Each state is held inside of a POJO where the key is the state name.

The WeakMaps would be set like so:

```js
WeakMapA.set(this.get('email'), new MapWithDefault());
WeakMapB.set(this.get('email'), new MapWithDefault());
```


Now when it comes to actually returning the value for the CP. This is what we would have to do:

```js
return computed(propertyName, function() {
  let propertyValue = this.get(propertyName);
  let stateWeakMap = stateHashMap[stateName];

  return stateWeakMap.get(propertyValue).get('XXX');
});
```

File: addon/custom-map-with-default.js (hunk 6 above)
Comment (thoov) [inline addon/custom-map-with-default.js]: :+1:

**Comment by thoov** [comment]
basically calling WeakMapA.get(this.get('email')) should return the state object and not a MapWithDefault Object that within it contains the state

**Comment by stefanpenner** [comment]
Does it seem like the MapWithDefault is superseeded by the WeakMap?

```
return computed(propertyName, function() {
  let propertyValue = this.get(propertyName);
  let stateWeakMap = stateHashMap[stateName];
  return stateWeakMap.get(propertyValue); // this feels sufficient
});

**Comment by thoov** [comment]
> Does it seem like the MapWithDefault is superseded by the WeakMap?

Exactly

**Comment by stefanpenner** [comment]
> Does it seem like the MapWithDefault is superseded by the WeakMap?
> Exactly

im not really sure why the MapWithDefault is still present

**Comment by thoov** [comment]
It wont be anymore after this next commit

**Comment by stefanpenner** [comment]
> It wont be anymore after this next commit
:fire:

**Comment by thoov** [comment]
@stefanpenner now take a look at: https://github.com/thoov/ember-state-services/blob/newExample/addon/state-for.js.

If that looks ok then:

- [ ] Add 'draft' indicator to example
- [ ] Add a few more test cases

**Comment by stefanpenner** [comment]
https://github.com/thoov/ember-state-services/blob/newExample/addon/state-for.js#L14 should be named  `weakMaps`, and we may want a way to purge a `weakMap`

**Comment by stefanpenner** [comment]
missing close parens https://github.com/thoov/ember-state-services/blob/newExample/addon/state-for.js#L21-L22

**Comment by stefanpenner** [comment]
should use the built-in `typeof`

- https://github.com/thoov/ember-state-services/blob/newExample/addon/state-for.js#L31
- https://github.com/thoov/ember-state-services/blob/newExample/addon/state-for.js#L64

**Comment by stefanpenner** [comment]
if initial state is set, but not a function. We should likely throw a typeError
https://github.com/thoov/ember-state-services/blob/newExample/addon/state-for.js#L66

**Comment by thoov** [comment]
@stefanpenner Just thinking off the top of my head for purging weakMap we could do one of the following:

```js
 stateFor('state-name').purge(); // or .clear()
```

```js
import { stateFor, deleteStateFor } ....
```

**Comment by thoov** [comment]
Im more of a fan of the first one

**Comment by stefanpenner** [comment]
Purge isn't really truthful, since it just destroys the polyfil weakmap. Which won't actually delete the related data. I'm inclined to punt on it for now.  

If we do want it, maybe release is the right word? As it does not actively purge anything, rather removes it's reference and if nothing else holds a referenced it is released

**Comment by thoov** [comment]
@stefanpenner I am on board with just punting on this for now. I also added some more unit tests and added the 'draft' indicator on the list.

**Comment by stefanpenner** [comment]
Nice. This lgtm. Merge when you feel good about it.

**Pull request merged**